### PR TITLE
Feature/selfhosted

### PR DESCRIPTION
**Pull request by wiardvanrij**
- It can use `VERSION_CHECKER_SELFHOSTED_TOKEN` for the token to talk to your registry (You can compare it to JWT)
- It can use basic auth with `VERSION_CHECKER_SELFHOSTED_USERNAME` and `VERSION_CHECKER_ SELFHOSTED_PASSWORD` to fetch a token. I do however recommend just using the token.
- As argument we can use `--selfhosted-registry-url`. This should be the host for your API endpoint: `https://repositories.yourdomain.com` - An extra path is allowed, such as `https://repositories.yourdomain.com/artifactory`
- It does not (yet) implement pagination in fetching all the tags. This simply does not work on every system and there are too many issues currently. This is now 'hard fixed' on n=500
- It uses the `/V2` API endpoint
- It uses `application/vnd.docker.distribution.manifest.list.v1+json` because this is still the default, it gives us the `created timestamp`. Also not every registry has v2.2 manifest.

**What is not implemented:**

- It does not refresh the auth token
- My experience is that most implementations have some form of unlimited token life-time, but it would be a 'nice to have' in the future
- Some form of delay between requests when fetching the manifests. This _could_ cause issues depending on the quality of your registry and the amount of images + tags present.

-- 
Relates to #30 and #14

### CODE REVIEW

**Comment by wiardvanrij** [comment]
/assign @joshvanl

I've reworked the URL part and included sane tests for this. Further information is in the first comment :)

File: pkg/client/selfhosted/selfhosted.go [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+	Created time.Time `json:"created,omitempty"`
+}
+
+func New(ctx context.Context, opts Options) (*Client, error) {
+	client := &http.Client{
+		Timeout: time.Second * 5,
+	}
+
+	// Initialize base vars for when no URL for this client is set.
+	var hostRegex *regexp.Regexp = nil
```
Comment (JoshVanL) [inline pkg/client/selfhosted/selfhosted.go:84]: No need for `=`, they are already initialized to nil.

File: pkg/client/selfhosted/selfhosted.go [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+	Name      string  `json:"name"`
+	Timestamp string  `json:"last_updated"`
+	Images    []Image `json:"images"`
+}
+
+type Image struct {
+	Digest       string `json:"digest"`
+	OS           string `json:"os"`
+	Architecture string `json:"Architecture"`
+}
```
Comment (JoshVanL) [inline pkg/client/selfhosted/selfhosted.go:62]: Used?

File: pkg/client/selfhosted/selfhosted.go [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+type TagResponse struct {
+	Name string   `json:"name"`
+	Tags []string `json:"tags"`
+}
+
+type Result struct {
+	Name      string  `json:"name"`
+	Timestamp string  `json:"last_updated"`
+	Images    []Image `json:"images"`
+}
```
Comment (JoshVanL) [inline pkg/client/selfhosted/selfhosted.go:56]: Used?

**Review by JoshVanL — CHANGES_REQUESTED** [review]
Hi @wiardvanrij, unfortunately it's still not quite right, I'm getting the wrong SHA back for image versions:

```
api.ImageTag{Tag:"v0.0.1", SHA:"sha256:6bd691eb01ed11e699f49390db1f2a1343b81745e7c1d000697f19176f4f82c6", Timestamp:time.Time{wall:0xb20d82b, ext:63734127029, loc:(*time.Location)(nil)}, Architecture:"amd64", OS:""}
api.ImageTag{Tag:"v0.1.1", SHA:"sha256:93863074a3ce57173d8f336c266d3b72b0e43919ba003110e63e7cf6321f0994", Timestamp:time.Time{wall:0xb681bb7, ext:63733975395, loc:(*time.Location)(nil)}, Architecture:"amd64", OS:""}


$ docker images --digests | grep 'localhost:5000/version-checker' | grep v0
localhost:5000/version-checker                                                   v0.0.1                                     sha256:64b8ec38181ba23e331eb450a081515b7e0e4e21cc92e1013122c22c856706a4   931c69673c83        53 minutes ago      50.8MB
localhost:5000/version-checker                                                   v0.1.1                                     sha256:40a1f8c4f8e1e45ded4c26c67531fe5e876513842327ef48867dface177f25c2   b803b1053a84        43 hours ago        54.6MB
```

```
	client, err := New(context.TODO(), Options{
		URL: "http://localhost:5000",
	})
	if err != nil {
		t.Fatal(err)
	}

	tags, err := client.Tags(context.TODO(), "", "", "version-checker")
	if err != nil {
		t.Fatal(err)
	}

	for _, tag := range tags {
		fmt.Printf("%#+v\n", tag)
	}
```

This issue is that we need to make two calls for each manifest with different headers.
1. `"Accept: application/vnd.docker.distribution.manifest.v1+json"`: This contains the created timestamp that we need.
2. `"Accept: application/vnd.docker.distribution.manifest.v2+json"`:  This contains the correct image SHA tag in the `Docker-Content-Digest` header.

Happy to pick this up, know it been going back and forth a few times :grimacing: 

/assign @wiardvanrij 
/unassign

**Comment by wiardvanrij** [comment]
@JoshVanL 

Ye it would be great if you can pick that up.  I've actually sort of found the problem with that. 

When I do a request to my registry with v2 request header, I receive the v2 body but a v1 header. I do wonder though what you can make of it, and if that would work for my system :)

**Comment by JoshVanL** [comment]
Hi @wiardvanrij, I've updated it a bit. Do let me know if that works for you

File: pkg/client/selfhosted/selfhosted.go [hunk 4]
```diff
@@ -76,105 +65,98 @@ type V1Compatibility struct {
 }
 
 func New(ctx context.Context, opts Options) (*Client, error) {
-	client := &http.Client{
-		Timeout: time.Second * 5,
+	client := &Client{
+		Client: &http.Client{
+			Timeout: time.Second * 5,
```
Comment (wiardvanrij) [inline pkg/client/selfhosted/selfhosted.go:70]: I've hit this timeout about 10-20% of the time. When I changed it to 10 seconds, I had 0 issues.

**Comment by wiardvanrij** [comment]
> Hi @wiardvanrij, I've updated it a bit. Do let me know if that works for you

@JoshVanL Really neat. It works great. I personally found some issues, but that's 100% an Artifactory issue. I was wondering the following use-case:

- Finds image
- Fetches tags
- Going to fetch all manifests from above tags
- What if 1 tag fails to get a manifest? 

I believe that it will re-queue and ignore the entire image? Because that's my case at the moment that 1 tag is "bugged out".  Would it be nice to use some "best-effort" for fetching image tags/manifests? 

Anyhow, thats again an edge case :)

**Comment by JoshVanL** [comment]
> > Hi @wiardvanrij, I've updated it a bit. Do let me know if that works for you
> 
> @JoshVanL Really neat. It works great. I personally found some issues, but that's 100% an Artifactory issue. I was wondering the following use-case though:
> 
> * Finds image
> * Fetches tags
> * Going to fetch all manifests from above tags
> * What if 1 tag fails to get a manifest?
> 
> I believe that it will re-queue and ignore the entire image? Because that's my case at the moment that 1 tag is "bugged out". Would it be nice to use some "best-effort" for fetching image tags/manifests?
> 
> Anyhow, thats again an edge case :)

Urgh, that's gross. Starting to hate registries... 

Is there any status code that we can take advantage of? A 404 could be helpful (we've been told the tag exists but the manifests doesn't, so we can safely ignore).

**Comment by wiardvanrij** [comment]
> > > Hi @wiardvanrij, I've updated it a bit. Do let me know if that works for you
> > 
> > 
> > @JoshVanL Really neat. It works great. I personally found some issues, but that's 100% an Artifactory issue. I was wondering the following use-case though:
> > 
> > * Finds image
> > * Fetches tags
> > * Going to fetch all manifests from above tags
> > * What if 1 tag fails to get a manifest?
> > 
> > I believe that it will re-queue and ignore the entire image? Because that's my case at the moment that 1 tag is "bugged out". Would it be nice to use some "best-effort" for fetching image tags/manifests?
> > Anyhow, thats again an edge case :)
> 
> Urgh, that's gross. Starting to hate registries...
> 
> Is there any status code that we can take advantage of? A 404 could be helpful (we've been told the tag exists but the manifests doesn't, so we can safely ignore).

Well, then you are going to like the fact that I'm getting a 403 because it thinks the image is out of scope lol. I shared your feeling about registries a few days ago already ;) 

Anyhow, 403 and the body is:

```
{
  "errors": [
    {
      "code": "UNAUTHORIZED",
      "message": "The client does not have permission for manifest: Download request for repo:path '{redacted}:{repo}/cluster-{image}/{tag}/manifest.json' is forbidden for user 'anonymous'.",
      "detail": {
        "manifest": "{repo}/{image}"
      }
    }
  ]
}
```

**Comment by JoshVanL** [comment]
@wiardvanrij I have updated it a bit, let me know if that works for you and we can merge :)

**Comment by wiardvanrij** [comment]
/lgtm @JoshVanL ! 

No errors or whatsoever and the results look good too. When there is a release I will roll this out on our TST env and keep it running for a while.

**Comment by JoshVanL** [comment]
@wiardvanrij Awesome, thanks! :smile: 

/lgtm

**Comment by wiardvanrij** [comment]
@JoshVanL Party time! Have a good weekend

**Comment by JoshVanL** [comment]
> @JoshVanL Party time! Have a good weekend

You too! :partying_face:

**Pull request merged**